patchwork-fml/src/main/java/net/minecraftforge/forgespi/language/ModFileScanData.java

```diff
@@ -0,0 +1,257 @@
+				Annotation annotationObject = getAnnotationObject(clazzObj, annotationType);
+
+				if (annotationObject == null) {
+					LOGGER.error(String.format("Cannot fetch annotation object %s %s %s %s",
+							annotationType, targetType, clazz, memberName
+					));
+					return;
+				}
+
+				Method[] argMethods = annotationObject.getClass().getDeclaredMethods();
+
+				for (Method argMethod : argMethods) {
+					if (isArgumentMethod(argMethod)) {
+						annotationData.put(
+								argMethod.getName(),
+								argMethod.invoke(annotationObject)
+						);
+					}
+				}
+			} catch (Throwable e) {
+				LOGGER.error(e);
+			}
+		}
+
+		private static boolean isArgumentMethod(Method method) {
+			String name = method.getName();
+			if (name.equals("toString")) return false;
+			if (name.equals("hashCode")) return false;
+			if (name.equals("getClass")) return false;
+			if (name.equals("equals")) return false;
+			if (name.equals("annotationType")) return false;
+
+			return true;
+		}
+
+		private Annotation getAnnotationObject(Class<?> clazzObj, Class annotationType) {
+			try {
+				switch (targetType) {
+				case TYPE:
+					return clazzObj.getAnnotation(annotationType);
+				case FIELD:
+					return clazzObj.getField(memberName)
+							.getAnnotation(annotationType);
+				case METHOD:
+					//TODO handle overloaded methods
+					String methodName = memberName.substring(0, memberName.indexOf('('));
+					return Arrays.stream(clazzObj.getDeclaredMethods())
+							.filter(method -> method.getName().equals(methodName))
+							.findFirst()
+							.orElseThrow(() -> new RuntimeException("Cannot get method " + memberName))
+							.getAnnotation(annotationType);
+				default:
+					return null;
+				}
+			} catch (NoSuchFieldException e) {
+				return null;
+			}
+		}
+
+		@Override
+		public boolean equals(final Object obj) {
+			if (!(obj instanceof AnnotationData)) {
+				return false;
+			}
+
+			AnnotationData dat = (AnnotationData) obj;
+			return Objects.equals(annotationType, dat.annotationType)
+					&& Objects.equals(targetType, dat.targetType)
+					&& Objects.equals(clazz, dat.clazz)
+					&& Objects.equals(memberName, dat.memberName);
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hash(annotationType, targetType, clazz, memberName);
+		}
+	}
+
+	//if it's a jij mod, return parent mod's container
+	private static ModContainer getModContainer(String modid) {
+		return FabricLoader.INSTANCE.getAllMods().stream().filter(modContainer -> {
+			LoaderModMetadata info = ((net.fabricmc.loader.ModContainer) modContainer).getInfo();
+			Collection<NestedJarEntry> jars = info.getJars();
+
+			if (modContainer.getMetadata().getId().equals(modid)) {
+				return false;
+			}
+
+			return jars.stream().anyMatch(
+					jar -> jar.getFile().contains(modid)
+			);
+		}).findAny().orElse(FabricLoader.INSTANCE.getModContainer(modid).orElseThrow(
+				() -> new RuntimeException("Cannot get mod container for " + modid)
+		));
+	}
```

Comment:
I really don't like this. How about we have a custom attribute that designates the "owner" modid for the annotation data?

Reply:
I found that there is "modmenu:parent" in fabric.mod.json. Using that could be appropriate